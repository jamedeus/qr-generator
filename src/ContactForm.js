import React, { useState } from 'react';
import Form from 'react-bootstrap/Form';
import FloatingLabel from 'react-bootstrap/FloatingLabel';
import Button from 'react-bootstrap/Button';


function ContactForm({ onSubmit }) {
    // Create form validation state object
    const [validated, setValidated] = useState(false);

    // Wrap submit handler, validate fields before calling
    const handleSubmit = (event) => {
        // Prevent submitting with invalid fields
        if (event.currentTarget.checkValidity() === false) {
            event.preventDefault();
            event.stopPropagation();
        // Pass event to submit handler if all fields valid
        } else {
            onSubmit(event);
        }

        // Show validation highlights
        setValidated(true);
    };

    // Format phone number as user types
    function formatPhone(event) {
        // Remove all non-numeric characters, 10 digits max
        const input = event.target.value.replace(/\D/g,'').substring(0,10);

        // Parse phone number sections
        const areaCode = input.substring(0,3);
        const middle = input.substring(3,6);
        const last = input.substring(6,10);

        // Format based on current length
        if (input.length > 6) {
            event.target.value = `(${areaCode}) ${middle}-${last}`;
        } else if (input.length > 3) {
            event.target.value = `(${areaCode}) ${middle}`;
        } else if (input.length > 0) {
            event.target.value = `(${areaCode}`;
        } else if (input.length == 0) {
            event.target.value = ``;
        }
    }

    // Allow all characters except spaces in email field
    function formatEmail(event) {
        if (event.code === 'Space') {
            event.preventDefault();
        }
    }

    return (
        <Form noValidate validated={validated} onSubmit={handleSubmit}>
            <FloatingLabel label="First Name" className="mb-3">
                <Form.Control type="text" name="firstName" placeholder="First Name" required />
            </FloatingLabel>

            <FloatingLabel label="Last Name" className="mb-3">
                <Form.Control type="text" name="lastName" placeholder="Last Name" required />
            </FloatingLabel>

            <FloatingLabel label="Email" className="mb-3">
                <Form.Control type="email" name="email" placeholder="Email" onKeyDown={formatEmail} required />
            </FloatingLabel>

            <FloatingLabel label="Phone" className="mb-3">
                <Form.Control type="tel" name="phone" placeholder="Phone" onChange={formatPhone} pattern="^(\d{10}|\(\d{3}\) \d{3}-\d{4})$" required />
            </FloatingLabel>

            <div className="d-flex">
                <Button variant="primary" type="submit" className="mx-auto">
                    Generate
                </Button>
            </div>
        </Form>
    )
}


export default ContactForm;
