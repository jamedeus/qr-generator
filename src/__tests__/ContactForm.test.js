import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import ContactForm from '../ContactForm';

describe('ContactForm', () => {
    let component, user;

    beforeEach(() => {
        // Render component, create userEvent instance
        user = userEvent.setup();
        component = render(
            <ContactForm />
        );
    });

    it('matches snapshot', () => {
        expect(component).toMatchSnapshot();
    });

    it('formats phone number correctly, blocks invalid characters', async () => {
        const phoneInput = component.getByPlaceholderText('Phone');

        // Should not accept non-numeric characters
        await user.type(phoneInput, 'asdf!@#$%^&*()-_=+');
        expect(phoneInput.value).toBe('');

        // Simulate user input, confirm each block correctly formatted
        await user.type(phoneInput, '123');
        expect(phoneInput.value).toBe('(123');
        await user.type(phoneInput, '456');
        expect(phoneInput.value).toBe('(123) 456');
        await user.type(phoneInput, '7890');
        expect(phoneInput.value).toBe('(123) 456-7890');

        // Simulate 10 backspaces, should automatically remove formatting characters
        const backspaces = '{backspace}'.repeat(10);
        await user.type(phoneInput, backspaces);
        expect(phoneInput.value).toBe('');
    });

    it('blocks spaces in email field', async () => {
        const emailInput = component.getByPlaceholderText('Email');
        await user.type(emailInput, 'test space@example.com');

        expect(emailInput.value).toBe('testspace@example.com');
    });

    it('requires all fields, correct email and phone format', async () => {
        // Get reference to each field
        const firstName = component.getByPlaceholderText('First Name');
        const lastName = component.getByPlaceholderText('Last Name');
        const emailInput = component.getByPlaceholderText('Email');
        const phoneInput = component.getByPlaceholderText('Phone');

        // All empty fields should be invalid
        expect(firstName.validity.valueMissing).toBe(true);
        expect(lastName.validity.valueMissing).toBe(true);
        expect(emailInput.validity.valueMissing).toBe(true);
        expect(phoneInput.validity.valueMissing).toBe(true);

        // Add arbitrary text to name fields, should be valid
        firstName.value = "foobar";
        expect(firstName.validity.valid).toBe(true);
        lastName.value = "foobar";
        expect(lastName.validity.valid).toBe(true);

        // Add arbitrary text to email, should fail type match
        emailInput.value = "foobar";
        expect(emailInput.validity.valid).toBe(false);
        expect(emailInput.validity.typeMismatch).toBe(true);

        // Add arbitrary text to phone, should fail pattern match
        phoneInput.value = "foobar";
        expect(phoneInput.validity.valid).toBe(false);
        expect(phoneInput.validity.patternMismatch).toBe(true);

        // Add correct email format, should be valid
        emailInput.value = "foo@bar.com";
        expect(emailInput.validity.valid).toBe(true);

        // Add correct phone format, should be valid
        phoneInput.value = "(123) 456-7890";
        expect(phoneInput.validity.valid).toBe(true);
    });
});
