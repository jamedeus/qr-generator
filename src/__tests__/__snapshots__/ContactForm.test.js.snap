// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm matches snapshot 1`] = `
<form
  className=""
  noValidate={true}
  onSubmit={[MockFunction]}
>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="firstName"
      placeholder="First Name"
      required={true}
      type="text"
    />
    <label>
      First Name
    </label>
  </div>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="lastName"
      placeholder="Last Name"
      required={true}
      type="text"
    />
    <label>
      Last Name
    </label>
  </div>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="email"
      onKeyDown={[Function]}
      placeholder="Email"
      required={true}
      type="email"
    />
    <label>
      Email
    </label>
  </div>
  <div
    className="mb-3 form-floating"
  >
    <input
      className="form-control"
      name="phone"
      onChange={[Function]}
      pattern="^(\\d{10}|\\(\\d{3}\\) \\d{3}-\\d{4})$"
      placeholder="Phone"
      required={true}
      type="tel"
    />
    <label>
      Phone
    </label>
  </div>
  <div
    className="d-flex"
  >
    <button
      className="mx-auto btn btn-primary"
      disabled={false}
      type="submit"
    >
      Generate
    </button>
  </div>
</form>
`;
